### DISCUSSION TITLE
[New] Utiliser le stockage temporaire et les agrégats

### DISCUSSION THREAD

**Discussion by DocRoms** [cnumr/best-practices#47] [Bonnes pratiques] [answered]
### Bonne pratique : Utiliser le stockage temporaire et les agrégats
Le stockage temporaire, ce n'est pas juste un fichier que l'on supprime après telle ou telle durée. De nos jours de nombreuses bases de données gèrent ce que l'on appelle le "`TTL`" (Time to Live : la durée de vie). 

Une donnée peut elle aussi avoir une durée de vie, il est même possible de l'agréger à une table d'agrégation lorsque celle-ci s'apprêtent à être détruite : il existe de nombreuses solutions pour optimiser nos bases de données et éviter qu'elles ne grossissent continuellement le long de la durée de vie de notre application.

par exemple `MongoDB` , `AWS DynamoDB`, `Redis`, `Oracle` et `Azure DocumentDB` gèrent très bien le TTL.

### Règle de validation proposée :
    - nombre de données agrégable n'ayant pas de TTL : 0

### Sources : 
[Azue et CosmoDb](https://docs.microsoft.com/fr-fr/azure/cosmos-db/mongodb/mongodb-time-to-live)
[MongoDB](https://docs.mongodb.com/manual/core/index-ttl/)
[Redis](https://redis.io/commands/TTL)
[Oracle](https://docs.oracle.com/cd/E17277_02/html/GettingStartedGuide/timetolive.html)

**Comment by arnolem** [comment]
En apparence, c'est une règle qui est très intéressante.

Mais dans la pratique, j'ai déjà eu 2 cas ou j'ai finalement fait des retours en arrière en supprimant des redis ou des influxdb à cause de la sur-consommation en CPU de ce type de solution.

Mettre en place un TTL sur une base de données déjà existante ne fait pas trop débat. Mais faire un choix différent d'architecture pour mettre en place ce type de base de données pourrait être contre-productif au niveau de la consommation CPU.

Je me base uniquement sur mon expérience sur deux projets, reste à voir les retours de chacun sur ce sujet.

**Reply by DocRoms** [reply]
C'est exact, c'est vrai qu'avec des bases énormes, qui ne sont pas architecturée correctement il peut y avoir des problèmes de CPUs, le TTL effectue un simple `DELETE` dans un `bgTask`, ce qui ajoute, de facto, une tâche de fond supplémentaire chargée d'effectuer le clean de temps en temps à son architecture (+ de CPU et de RAM utilisée pour elle).

Le clean est fait de manière régulière mais n'est pas garantis effectif à la date <t>, généralement la tâche de fond s'exécute toutes les minutes je crois (sur Mango) on peut donc avoir une suppression à + une minute par rapport au TTL demandé, ce n'est pas quelques choses de plus "lourd" qui s'effectuent instantanément. Ca peut-être très utile dans de nombreux cas, mais effectivement, ce n'est pas quelques chose à activer sur toutes les tables, partout. Je penses qu'il y a un vrai intérêt concernant la rétention de logs users, ce qui évite d'avoir des milliards de lignes de connexions, et ce qui permet d'être un peu plus adaptés aux recommandations RGPD, mais je te rejoins sur le fait de trouver le juste milieu entre le sur-stockage que l'on peut avoir dans une base, et la tâche de fond qui tourne en permanence pour supprimer les données expirées.

Je suis pressé de lire les autres avis sur la question.

**Comment by tbroyer** [comment]
En fait je ne comprend pas la règle telle que formulée ici.

Elle semble aussi parler de deux choses bien distinctes finalement (par agrégats, on parle finalement de time series ici, non ?), et en se concentrant sur l'aspect technique/technologique plutôt que sur l'objectif recherché.

Si l'objectif est le surstockage, alors oui avoir une routine de purge (lancée toutes les nuits / weekends / mois), ou tirer partie des indexes TTL des bases de données peut être une solution (la purge n'est d'ailleurs pas forcément basée sur le temps qui passe, et/ou dépendante d'autres données : par exemple après une suppression logique).
L'utilisation de time series, qui perdent en granularité avec le temps, peut également être une solution, mais les implications fonctionnelles (et aussi technique d'ailleurs) sont quand même assez différentes.

En terme d'énoncé de règle, je parlerai donc plutôt de purger les données qui ne sont plus nécessaires (avec les solutions techniques mentionnées), voire (mais ça pourrait presque être une règle distincte) de modifier leur stockage avec perte de granularité d'information (de préférence en choisissant une base de données qui gère directement les time series, plutôt que les implémenter soi-même)

**Comment by rlemaire** [comment]
Je pense qu'on pourrait généraliser la règle avec quelque chose comme « Avoir une politique d'expiration et de suppression des données ».
L'implémentation technique TTL ou batch est secondaire, dépend du contexte, et peut être proposé en exemple dans la description.

**Reply by ACTLEM** [reply]
Je suis plutôt pour. Je pose quand même la question de la règle de validation de cette recommandation, comment en avoir une efficace et pertinente? Une idée à chaud: "Nombre de données inutilisées depuis x temps(?) et stockées et sans contrainte légale <1"

**Reply by ldesgrange** [reply]
> « Avoir une politique d'expiration et de suppression des données »

+1 et rappeler que c'est une obligation légale en ce qui concerne les données personnelles (RGPD).

**Comment by ACTLEM** [comment, marked as answer]
Note équipe projet: Remplacer la BP 68 (cf https://github.com/cnumr/best-practices/discussions/80) via une bonne pratique sur la politique d'expiration et suppression des données

Issue créée: https://github.com/cnumr/best-practices/issues/176
